Sources/SwiftGRPCNIO/GRPCServerCodec.swift

```diff
@@ -0,0 +1,63 @@
+import Foundation
+import SwiftProtobuf
+import NIO
+import NIOHTTP1
+
+public enum GRPCServerRequestPart<MessageType: Message> {
+  case headers(HTTPRequestHead)
+  case message(MessageType)
+  case end
+}
+
+public enum GRPCServerResponsePart<MessageType: Message> {
+  case headers(HTTPHeaders)
+  case message(MessageType)
+  case trailers(HTTPHeaders)
+}
+
+/// A simple channel handler that translates raw gRPC packets into decoded protobuf messages,
+/// and vice versa.
+/// **Currently unused, as we do not yet know the request's method name (and thus message types) when this object is instantiated.**
```

Comment:
You can work around this by having this handler be a handler-factory: that is, this handler catches the initial inbound messages and, based on the request's method name, generates an appropriate handler and inserts it in the pipeline directly behind itself.

Does that make sense?

Reply:
That would definitely make sense; it's essentially what I had tried to in a much more clunky fashion by piping messages to the `GRPCChannelHandler`. Is there an example on how that insertion process would be done correctly?

Reply:
Not directly, but there is some similar stuff in the `HTTPUpgradeHandler`: https://github.com/apple/swift-nio/blob/a7fee8d98ffb1c1fd695a92521c1f6e59a98c78b/Sources/NIOHTTP1/HTTPUpgradeHandler.swift#L212-L242

The general idea is that, in general, pipeline manipulating operations on the channel you're currently running on actually complete synchronously. You shouldn't rely on that fact (that is, you should be prepared to buffer data), but you can always assume any requirement to buffer data will be short-lived.

The only thing you're missing is addinghandlers. For that you probably want [`add(handler:after:)`](https://apple.github.io/swift-nio/docs/current/NIO/Classes/ChannelPipeline.html#/s:3NIO15ChannelPipelineC3addAA15EventLoopFutureCyytGSSSg4name_AA0B7Handler_p7handlerAaJ_p5aftertF), where `after` can be `self`.

Reply:
Thanks! I have changed the code to directly add the `ServerCodec` and `CallHandlers` now and remove the `ChannelHandler`.